File: Tools/BezierCurve.py
import scipy
import bisect
import numpy

class BezierCurve:
    ''' Calculates a Bezier Spline, like in the Color-Curve-Dialog in Gimp
        The Cubic Bezier-Curve is: y=P0*(1-t)**3+3*P1*(1-t)**2*t+3*P2*(1-t)*t**2+P3*t**3
        This gives with t=(x - x_i)/(x_i+1 - x_i), dx=x - x_i, D=x_i+1 - x_i
        C0=P0
        C1=(3*P1-3*P0)/D
        C2=(3*P3-6*P1+3*P0)/D**2        
        C3=(P3-3*P2+3*P1-P0)/D**3
        
        y=(((C3*dx)+C2)*dx+C1)*dx+C0
        
        and again with
        
        D0=C0 - C1*x_i + C2*x_i**2 - C3*x_i**3
        D1=C1 -2*C2*x_i + 3*C3*x_i**2
        D2=C2 - 3*C3*x_i
        D3=C3
        
        y=(((D3*x)+D2)*x+D1)*x+D0
    '''
    def __init__(self,  x,  y):
        # this ist stolen from GIMP. Hale the GPL!!!
        # File: GIMP/app/base/curves.c
        self.x=x
        if len(x)<2:
            raise ValueError,  'need minimal 2 Points'
        elif len(x)==2:
            self.P=[(y[0],  y[0]+(y[1]-y[0])/3.0,  y[0]+2.0*(y[1]-y[0])/3.0,  y[1])]
            self.C=[(y[0],  (y[1]-y[0])/(x[1]-x[0]),  0.0,  0.0)]
            self.D=[((y[0]-(y[1]-y[0])/(x[1]-x[0])*x[0]),  (y[1]-y[0])/(x[1]-x[0]),  0.0,  0.0)]
        else:
            self.P=[]
            self.C=[]
            self.D=[]
            for i in range(len(x)-1):
                P0=y[i]
                P3=y[i+1]
                dx=x[i+1]-x[i]
                if i==0:
                    #only the right neighbor is available. Make the tangent at the
                    #right endpoint parallel to the line between the left endpoint
                    #and the right neighbor. Then point the tangent at the left towards
                    #the control handle of the right tangent, to ensure that the curve
                    #does not have an inflection point.
                    slope = (y[i+2] - y[i]) / (x[i+2] - x[i]);
                    P2 = y[i+1] - slope * dx / 3.0;
                    P1 = y[i] + (P2 - y[i]) / 2.0;
                elif i==len(x)-2:
                    slope = (y[i+1] - y[i-1]) / (x[i+1] - x[i-1]);
                    P1 = y[i] + slope * dx / 3.0;
                    P2 = y[i+1] + (P1 - y[i+1]) / 2.0;
                else:
                    # Both neighbors are available. Make the tangents at the endpoints
                    # parallel to the line between the opposite endpoint and the adjacent
                    #neighbor.
                    slope = (y[i+1] - y[i-1]) / (x[i+1] - x[i-1])
                    P1 = y[i] + slope * dx / 3.0
                    slope = (y[i+2] - y[i]) / (x[i+2] - x[i])
                    P2 = y[i+1] - slope * dx / 3.0
                self.P.append((P0,  P1,  P2,  P3))
                
                C0=P0
                C1=(        3*P1-3*P0)/dx
                C2=(   3*P2-6*P1+3*P0)/dx**2        
                C3=(P3-3*P2+3*P1-  P0)/dx**3
                self.C.append((C0, C1, C2, C3))
                
                D0=C0 -   C1*x[i] +   C2*x[i]**2 - C3*x[i]**3
                D1=C1 - 2*C2*x[i] + 3*C3*x[i]**2
                D2=C2 - 3*C3*x[i]
                D3=C3
                self.D.append((D0,  D1,  D2,  D3))

    def calcFromP(self,  x):
        n=min(max(bisect.bisect(self.x,  x)-1, 0),  len(self.x)-2)
        t=(x-self.x[n])/(self.x[n+1]-self.x[n])
        t1=1.0-t
        return self.P[n][0]*t1**3+3*self.P[n][1]*t1**2*t+3*self.P[n][2]*t1*t**2+self.P[n][3]*t**3

    def calcFromC(self,  x):
        n=min(max(bisect.bisect(self.x,  x)-1, 0),  len(self.x)-2)
        return ((self.C[n][3]*(x-self.x[n])+self.C[n][2])*(x-self.x[n])+self.C[n][1])*(x-self.x[n])+self.C[n][0]
    
    def calcFromD(self,  x):
        n=min(max(bisect.bisect(self.x,  x)-1, 0),  len(self.x)-2)
        return ((self.D[n][3]*x+self.D[n][2])*x+self.D[n][1])*x+self.D[n][0]

    def calcFromUnsortedScipyArr(self,  arr):
        sortedArr=scipy.array(arr)
        sortedArr.sort()
        nl=0
        for x,D in zip(self.x[1:],  self.D):
            D0, D1, D2, D3=D
            nr=bisect.bisect(sortedArr,  x)
            v=sortedArr[nl:nr]
            sortedArr[nl:nr]=((D3*v+D2)*v+D1)*v+D0
            nl=nr
        res=scipy.zeros(arr.shape)
        for i, n in enumerate(arr.argsort()):
            res[n]=sortedArr[i]
        return res.clip(0, 1)
        
    def __call__(self,  x):
        if type(x) in (int,  float):
            return self.calcFromP(x)
        elif type(x)==numpy.ndarray:
            return scipy.array(map(self.calcFromD,  x))
        elif hasattr(x,  '__iter__'):
            return map(self.calcFromD,  x)

File: ToolBox/CSource/BezierCurve.cpp
#include <BezierCurve.h>
#include <iostream>
#include <cmath>

using namespace std;

BezierCurve::BezierCurve(): params() {
    QList<QPointF> p;
    p.append(QPointF(0,0));
    p.append(QPointF(1,1));
    setPoints(p);
};

bool BezierCurve::setPoints(QList<QPointF>& p) {
    points=p;
    // this ist stolen from GIMP. Hale the GPL!!!
    // File: GIMP/app/base/curves.c
    params.clear();
    if (p.size()<2) {
        return false;
    } else if (p.size()==2) {
        CurveParams cp;
        cp.Xmin=-INFINITY;
        cp.Xmax=INFINITY;
        double dx=p.at(1).x()-p.at(0).x();
        if (dx==0.0)
            return false;
        double slope=(p.at(1).y()-p.at(0).y())/dx;
        cp.D0=p.at(0).y()-slope*p.at(0).x();
        cp.D1=slope;
        cp.D2=0.0;
        cp.D3=0.0;
        params.append(cp);
    } else {
        for (unsigned int i=p.size()-1; i--; ) {
            double p0=p.at(i).y();
            double p3=p.at(i+1).y();
            double p1;
            double p2;
            double dx=p.at(i+1).x()-p.at(i).x();
            if (dx==0.0)
                continue;
            
            if (i==0) {
                //only the right neighbor is available. Make the tangent at the
                //right endpoint parallel to the line between the left endpoint
                //and the right neighbor. Then point the tangent at the left towards
                //the control handle of the right tangent, to ensure that the curve
                //does not have an inflection point.
                double slope = (p.at(i+2).y() - p.at(i).y()) / (p.at(i+2).x() - p.at(i).x());
                p2 = p.at(i+1).y() - slope * dx / 3.0;
                p1 = p.at(i).y() + (p2 - p.at(i).y()) / 2.0;
            } else if (i==p.size()-2) {
                double slope = (p.at(i+1).y() - p.at(i-1).y()) / (p.at(i+1).x() - p.at(i-1).x());
                p1 = p.at(i).y() + slope * dx / 3.0;
                p2 = p.at(i+1).y() + (p1 - p.at(i+1).y()) / 2.0;
            } else {
                // Both neighbors are available. Make the tangents at the endpoints
                // parallel to the line between the opposite endpoint and the adjacent
                //neighbor.               
                double slope = (p.at(i+1).y() - p.at(i-1).y()) / (p.at(i+1).x() - p.at(i-1).x());
                p1 = p.at(i).y() + slope * dx / 3.0;
                slope = (p.at(i+2).y() - p.at(i).y()) / (p.at(i+2).x() - p.at(i).x());
                p2 = p.at(i+1).y() - slope * dx / 3.0;
            }
            double C0=p0;
            double C1=(        3*p1-3*p0)/dx;
            double C2=(   3*p2-6*p1+3*p0)/dx/dx;
            double C3=(p3-3*p2+3*p1-  p0)/dx/dx/dx;
            
            double x=p.at(i).x();
            CurveParams cp;
            
            cp.D0 = ((-C3*x +C2)*x-C1)*x+C0;
            cp.D1 = (3.0*C3*x -2.0*C2)*x+C1;
            cp.D2=C2 - 3.0*C3*x;
            cp.D3=C3;
            cp.Xmin=p.at(i).x();
            cp.Xmax=p.at(i+1).x();
            
            
            params.prepend(cp);
        }
        params.first().Xmin=-INFINITY;
        params.last().Xmax=INFINITY;
        return not params.empty();
    }
    return true;
}

QList<QPointF> BezierCurve::getPoints() {
    return points;
}

double BezierCurve::CurveParams::calc(double x) {
    return qMax(0.0, qMin(1.0, ((D3*x+D2)*x+D1)*x+D0));
}

double BezierCurve::operator()(double x) {
    if (params.empty())
        return 0.0;
    
    unsigned int p=getCurveParamIdx(x);
    return params[p].calc(x);
}

double BezierCurve::operator()(double x, unsigned int& hint) {
    while (params[hint].Xmax<x) hint++;
    while (params[hint].Xmin>x) hint--;
    return params[hint].calc(x);
}


QList<double> BezierCurve::range(double x0, double dx, unsigned int N) {
    QList<double> r;
    double x=x0;
    unsigned int p=getCurveParamIdx(x);
    unsigned int n=N;
    while (n) {
        CurveParams& cp=params[p];
        while (n && (x<cp.Xmax)) {
            r.append(cp.calc(x));
            x+=dx;
            n--;
        }
        p++;
    }
    return r;
}

QList<QPointF> BezierCurve::pointRange(double x0, double dx, unsigned int N) {
    QList<QPointF> r;
    double x=x0;
    unsigned int p=getCurveParamIdx(x);
    unsigned int n=N;
    while (n) {
        CurveParams& cp=params[p];
        while (n && (x<cp.Xmax)) {
            r.append(QPointF(x,cp.calc(x)));
            x+=dx;
            n--;
        }
        p++;
    }
    return r;
}


QList<double> BezierCurve::map(QList<double> X) {
    QList<double> r;
    for (unsigned int n=X.size(); n--; ) {
        double x=X[n];
        unsigned int p=getCurveParamIdx(x);
        r.prepend(params[p].calc(x));
    }
    return r;
}

QList<double> BezierCurve::mapSorted(QList<double> X) {
    QList<double> r;
    unsigned int p=getCurveParamIdx(X[0]);
    unsigned int n=X.size();
    while (n) {
        CurveParams& cp=params[p];
        double x;
        while (n && ((x=X[n-1])<cp.Xmax)) {
            r.prepend(cp.calc(x));
            n--;
        }
        p++;
    }
    return r;
}

QList<double> BezierCurve::mapSorted(QList<double> X, QList<unsigned int> Idx) {
    QList<double> r(X);
    unsigned int p=getCurveParamIdx(X[Idx[0]]);
    unsigned int n=X.size();
    while (n) {
        CurveParams& cp=params[p];
        double x;
        unsigned int idx=Idx[n-1];
        while (n && ((x=X[idx])<cp.Xmax)) {
            r[idx]=cp.calc(x);
            n--;
        }
        p++;
    }
    return r;
    
}

unsigned int BezierCurve::getCurveParamIdx(double x) {
    CurveParams cp;
    cp.Xmax=x;
    QList<CurveParams>::const_iterator iter=qLowerBound(params.constBegin(), params.constEnd(), cp);
    if (iter==params.constEnd())
        iter--;
    //cout << "iterpos " << iter-params.constBegin() << endl;
    return qMax(0,iter-params.constBegin());    
}

BezierCurve::CurveParams BezierCurve::getCurveParam(double x) {
    //cout << "get Curve Param for " << x << endl;
    unsigned int p=getCurveParamIdx(x);
    //cout << "is Param " << p << "/" << params.size() << endl;
    return params[p];
}


